llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -20241,6 +20241,10 @@ static SDValue trySimplifySrlAddToRshrnb(SDValue Srl, SelectionDAG &DAG,
     return SDValue();
   unsigned ShiftValue = SrlOp1->getZExtValue();
 
+  uint64_t EltSize = ResVT.getScalarSizeInBits();
+  if (ShiftValue > EltSize)
```

Comment:
I discussed this with @paulwalker-arm offline and we don't think this is right after all. I think we should just bail out of the optimisation if the ShiftValue exceeds the truncated element width. Also, it would be good to at least add an assert here that the shift value is zero. From what you're saying it should never happen, which makes sense.

Reply:
Can you please explain why you both came to the conclusion that this isn't correct?

Reply:
Sorry Matt this was my fault. My question arose from the typically behaviour of right shifts that are greater than the element bit length, whereby `i32 >> #32+N == i32 >> #32`.

Dave pointed out that in this case the element type in question is actually half sized because the combine is only expected to be called when a truncate is in play. That would imply `i32 >> 16+N == i32 >> 16`, which is clearly bogus.

Reply:
Thanks. I've now just gone with
```c++
  if (ShiftValue < 1 || ShiftValue > ResVT.getScalarSizeInBits())
    return SDValue();
```